### ISSUE TITLE
Add memoization for hooks

### ISSUE THREAD

**Issue by andyrichardson** [jaredpalmer/formik#1744]
## 🐛 Bug report

Hey there, thanks for the awesome library! Here are some thoughts on the hooks implementation and some challenges I've experienced.

Effects are intended to be triggered on dependency updates. React memoizes its setState functions to prevent infinite looping inside of effects. In the case where the existing state is required inside an effect, setState can be passed a function.

### Current Behavior

Hooks dependent on setState are infinitely re-created.

### Expected behavior

Hooks dependent on setState are memoised.

### Reproducible example

```js
// A: Not infinitely looping
useEffect(() => {
  setState(val => val + 1);
}, [setState]);

// B: Infinitely looping 
useEffect(() => {
  setState(state + 1);
}, [setState, state]);

// Formik specific
const { setValues } = useFormik();

// Infinitely looping as setValues is re-created on state change
// Also not possible to set state using existing state like with example A
useEffect(() => {
  setValues({});
}, [setValues]);

```

### Suggested solution(s)

Migrate any setters to not be dependent on state (only dependencies should be other setters). If reducers are in use, setting actions should take payloads which are either values, or functions which take existing values and return new values.

### Additional context

I'm happy to do some work on this but it would be a breaking change as validation needs to be reactive (i.e. validate on state change with useEffect rather than inside the setState function).

Here's [an example repro](https://github.com/jaredpalmer/formik/compare/master...andyrichardson:reactive-state-changes) for the setValues function

**Comment by johnrom** [comment]
React guarantees that the useState callback never changes, as shown in [the note below this section](https://reactjs.org/docs/hooks-reference.html#usestate). Therefore adding `setState` to any dependency is meaningless. I only know because I initially made a comment about this and had to correct myself :P. All setters need to be dependent on state because they need to know the current value of the state in order to set the new state. I'll use your example here:

```jsx
// B: Infinitely looping 
useEffect(() => {
  setState(state + 1);
}, [setState, state]);
```

If you do not include `state` in your dependencies, you are not going to infinitely loop. That is not a good thing, because this function as defined _is_ an infinite loop. The code says "every time the state changes, add 1 to the state". Adding 1 to the state changes the state, thus 1 should be added to _the new state_. If an infinite loop is not intended, then there must be some other dependency that is needed.

Finally, `useFormik` is not the right hook to use to access `setValues`. You're looking for `useFormikContext`. Refs: #1144 #1722 

I'm not saying you haven't found a bug, but I can't find one in the code you've presented as I haven't found any issues with infinite looping that weren't my own bugs. Please create a code sandbox based off of v2 with a concise reproduction of your issue.

v2 sandbox template: https://codesandbox.io/s/9w3zyjlqo

**Comment by andyrichardson** [comment]
@johnrom I think you're misunderstanding the issue. You're totally right, the setState function shouldn't ever change and therefore shouldn't need to be specified as a dependency.

If you [look at the example repro](https://github.com/jaredpalmer/formik/compare/master...andyrichardson:reactive-state-changes#diff-18d1f112a814aca08509a74bded25fbfL466) I shared, you'll see that the setState provided by formik does in fact change - this is the problem.

> Therefore adding setState to any dependency is meaningless

You'll always want to provide any "non React" setter function as a dependency as the implementation details are unknown. In this case, when you call set state without the dependency, you lose all your form state as the setState is called with stale stateful data.

> Finally, useFormik is not the right hook to use to access setValues. You're looking for useFormikContext

They are both return a Formik "client" - when you decide to use one or the other depends on context.

Anyhow, this is going way off topic - if you're not sure about what the problem is, I suggest you [take a look at the diff](https://github.com/jaredpalmer/formik/compare/master...andyrichardson:reactive-state-changes#diff-18d1f112a814aca08509a74bded25fbfL466) and ask yourself:
 - What happens when an old version of setValues is called
 - When will the setValue function be re-created
 - What happens if I make an effect that is dependent on both the state and the setter function
 - What happens if I make an effect that is dependent solely on the setter function, and calls said function

**Comment by johnrom** [comment]
> They are both return a Formik "client"

No they don't. `useFormik` is strictly for creating the internal tools used by Formik to initialize the Formik Context, which is why I've opened an issue suggesting we rename it #1722 . You can create your own context, sure. But you have to initialize that context using `FormikProvider` or a wrapper. If you are, awesome, but I write this for anyone who might stumble across and think "oh I'll just use that".

I'm sadly not familiar enough with the evolving v2 codebase to immediately deduce what you're trying to accomplish in your diff. Please reproduce your problem in a codesandbox. Here's my attempt. Note that when I add `formik.values` to either method, there's an infinite loop. That's expected, because the loop modifies `values` which the effect is dependent on. If I'm just dependent on `setValues`, however, there is no loop in either case. My guess is that either a recent update has fixed this issue, or there is something re-creating the component in your case which calls `useFormik` which is not allowing React to detect component identity.

If this isn't what you're trying to do, just fork the sandbox and show what the exact error is.

https://codesandbox.io/s/useformik-and-useformikcontext-3jncs

**Comment by andyrichardson** [comment]
> If I'm just dependent on setValues, however, there is no loop in either case. 

This wasn't the case in August - it looks like there has been a fix. 

There hasn't been a fix for the `setFormikState` function. PR has been updated again with a fix.

**Comment by andyrichardson** [comment]
Issue fixed by #1820.
